Additional Diagnostics in Dependency Injection

eerhardt opened the pull request:
Log events when a ServiceProvider is created:

* How many singleton, scoped, transient services?
* Log the list of registrations

Fix #56313

Of the list in #56313, I left out `How many container instances are in the processes?` because this can be determined from the number of `ServiceProviderBuilt` events that have occurred in the process.

eerhardt committed:
Additional Diagnostics in Dependency Injection

Log events when a ServiceProvider is created:

* How many singleton, scoped, transient services?
* Log the list of registrations

Fix #56313

eerhardt committed:
Add ServiceProvider HashCode to all events.

Fix ILLink warnings

josalem commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs:
```diff
@@ -62,68 +63,172 @@ public void ScopeDisposed(int serviceProviderHashCode, int scopedServicesResolve
+            }
+        }
+
+        private static void AppendServiceDescriptor(StringBuilder builder, ServiceDescriptor descriptor)
```
```suggestion
		[NonEvent]
        private static void AppendServiceDescriptor(StringBuilder builder, ServiceDescriptor descriptor)
```

davidfowl commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs:
```diff
@@ -62,68 +63,172 @@ public void ScopeDisposed(int serviceProviderHashCode, int scopedServicesResolve
+        [NonEvent]
+        public void ServiceProviderBuilt(ServiceProvider provider)
+        {
+            if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
```
I'd like to be able to have this fire OnEventCommand if I use a specific keyword so we can get access to this data even if the application is already built.

Here's what I did in Kestrel https://github.com/dotnet/aspnetcore/blob/545d4eece7fcc5a67b27f6ba96636d9f476741ae/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L354. The idea here is to handle when the event source is enabled and then fire the event (if the given keyword is on). In the kestrel case it always fires the event today because it doesn't have that much data (maybe we should tweak this), but for DI, we'd need to capture *all* service providers built in the application and then replay them here (and probably drop them from being rooted from the event source on dispose).

eerhardt committed:
PR feedback

eerhardt committed:
Write ServiceProvider information when DependencyInjectionEventSource becomes enabled.

This allows for listeners to attach to a process after it is running, and get the DI information.

eerhardt committed:
Update new events to use Informational level and to have a Keyword.

eerhardt commented:
@brianrob @noahfalk @josalem - can you give this a quick once over to ensure I'm using EventSource correctly?

josalem commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs:
```diff
@@ -14,9 +16,16 @@ internal sealed class DependencyInjectionEventSource : EventSource
+        // Event source doesn't support large payloads so we chunk large payloads like formatted call site tree and descriptors
         private const int MaxChunkSize = 10 * 1024;
 
+        private readonly List<ServiceProvider> _providers = new();
```
Do you care if this roots the `ServiceProvider`s until `DependencyInjectionEventSource.Log` gets cleaned up? That will happen in `AppContext.OnProcessEnd`. You could use a `WeakReference`, otherwise.

davidfowl replied:
Hmm maybe we should use a weak reference, that is a good idea.

brianrob replied:
If that's the case, you'll need to build up the state if you want to ensure that you can log it at anytime while the process is up, but yes, good point.

eerhardt replied:
Done.

> you'll need to build up the state if you want to ensure that you can log it at anytime while the process is up

If someone isn't `Dispose`'ing the ServiceProvider, and it gets garbage collected, I don't believe we want to log those providers if someone attaches after the provider has been collected. We really only care about "active" ServiceProviders.

josalem reviewed (commented):
A couple comments, but the `EventSource` end of things looks okay to me after a quick once-over.

brianrob commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs:
```diff
@@ -62,52 +71,195 @@ public void ScopeDisposed(int serviceProviderHashCode, int scopedServicesResolve
+                int scopedServices = 0;
+                int transientServices = 0;
+
+                StringBuilder descriptorBuilder = new StringBuilder("{ \"descriptors\":[ ");
```
Generally speaking, the pattern is to do less string formatting and to do more writing of events with payloads that can then be processed out-of-proc.  This looks like it might be a place where building up the state in-proc is preferred, or that you have to string-ify things anyway, but just wanted to call that out.  Definitely good that this is under its own keyword.

eerhardt replied:
Yeah, here we want to dump a bunch of data - the list of registered DI services. We are doing something similar above already:

https://github.com/dotnet/runtime/blob/a1aeb132ca60becb1e8ffe02150d59782c1214c2/src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs#L80-L94

brianrob commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs:
```diff
@@ -14,9 +16,16 @@ internal sealed class DependencyInjectionEventSource : EventSource
-        // Event source doesn't support large payloads so we chunk formatted call site tree
+        public static class Keywords
+        {
+            public const EventKeywords ServiceProviderInitialized = (EventKeywords)0x1;
```
I think you might need to specify a keyword for the other events in this provider now.  If you turn on keyword 0x1, you'll just get the ServiceProviderInitialized events.  If you specify keyword 0, I'm not sure you'll get the others - worth checking via ETW.

eerhardt replied:
>  If you turn on keyword 0x1, you'll just get the ServiceProviderInitialized events.

That was my intention. The existing events all check for `EventKeywords.All`:

https://github.com/dotnet/runtime/blob/a1aeb132ca60becb1e8ffe02150d59782c1214c2/src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs#L80-L82

So if you don't specify `All`, you won't get those events.

> If you specify keyword 0, I'm not sure you'll get the others - worth checking via ETW.

When I do

```
dotnet trace collect --providers Microsoft-Extensions-DependencyInjection:0:Verbose --name OrchardCoreTest
```

I don't get any events logged.

When I do

```
 dotnet trace collect --providers Microsoft-Extensions-DependencyInjection::Verbose --name OrchardCoreTest
```

I get them all.

And when I do

```
dotnet trace collect --providers Microsoft-Extensions-DependencyInjection:1:Verbose --name OrchardCoreTest
```

I just get the 2 new ones.

eerhardt committed:
Switch to use WeakReference when holding on to ServiceProviders in DependencyInjectionEventSource.

eerhardt committed:
Fix tests

eerhardt commented:
@davidfowl @maryamariyan - any other feedback here? I think this is ready to merge.

eerhardt commented:
`runtime-staging` failures are unrelated. Merging
